File: 백준/Gold/2357. 최솟값과 최댓값/최솟값과 최댓값.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    static class Value {
        int min;
        int max;

        public Value(int min, int max) {
            this.min = min;
            this.max = max;
        }
    }

    static class SegmentTree {
        Value[] segmentTree, arr;
        int leafSize, left, right;

        SegmentTree(int initSize) {
            leafSize = initSize;
            int height = (int) Math.ceil(Math.log(initSize) / Math.log(2));
            int treeSize = 2 << height;
            arr = new Value[initSize];
            segmentTree = new Value[treeSize];
            Arrays.fill(segmentTree, new Value(Integer.MAX_VALUE, Integer.MIN_VALUE));
        }

        void init(int node, int start, int end) {
            if (start == end) {
                segmentTree[node] = new Value(arr[start].min, arr[start].max);
                return;
            }
            int mid = (start + end) / 2;
            init(node * 2, start, mid);
            init(node * 2 + 1, mid + 1, end);
            int minValue = Math.min(segmentTree[node * 2].min, segmentTree[node * 2 + 1].min);
            int maxValue = Math.max(segmentTree[node * 2].max, segmentTree[node * 2 + 1].max);
            segmentTree[node] = new Value(minValue, maxValue);
        }

        Value startQuery(int left, int right) {
            this.left = left;
            this.right = right;
            return query(1, 0, leafSize - 1);
        }

        Value query(int node, int start, int end) {
            if (left > end || right < start) {
                return new Value(Integer.MAX_VALUE, Integer.MIN_VALUE);
            }
            if (left <= start && end <= right) {
                return segmentTree[node];
            }
            int mid = (start + end) / 2;
            Value leftValue = query(node * 2, start, mid);
            Value rightValue = query(node * 2 + 1, mid + 1, end);
            int minValue = Math.min(leftValue.min, rightValue.min);
            int maxValue = Math.max(leftValue.max, rightValue.max);
            return new Value(minValue, maxValue);
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        SegmentTree segmentTree = new SegmentTree(n);
        for (int i = 0; i < n; i++) {
            int tmp = Integer.parseInt(br.readLine());
            segmentTree.arr[i] = new Value(tmp, tmp);
        }
        segmentTree.init(1, 0, n - 1);
        StringBuilder sb = new StringBuilder();
        while (m-- > 0) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            Value tmp = segmentTree.startQuery(a - 1, b - 1);
            sb.append(tmp.min).append(" ").append(tmp.max).append("\n");
        }
        System.out.print(sb);
    }
}

File: 백준/Gold/2357. 최솟값과 최댓값/최솟값과 최댓값.cc
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

struct pt {
    ll mn, mx;
};

struct Seg {
    vector<pt> seg, arr;
    ll leafSize, l, r;

    Seg(int initSize) {
        leafSize = initSize;
        int h = ceil(log2(initSize));
        int sz = 2 << h;
        arr.resize(initSize);
        seg.resize(sz, {(ll) 9e18, (ll) -9e18});
    }

    void init(int n, int s, int e) {
        if (s == e) {
            seg[n] = arr[s];
            return;
        }
        int mid = (s + e) / 2;
        init(n * 2, s, mid);
        init(n * 2 + 1, mid + 1, e);
        seg[n].mn = min(seg[n * 2].mn, seg[n * 2 + 1].mn);
        seg[n].mx = max(seg[n * 2].mx, seg[n * 2 + 1].mx);
    }

    pt segQuery(int left, int right) {
        l = left;
        r = right;

        return query(1, 0, leafSize - 1);
    }

    pt query(int n, int s, int e) {
        if (l > e || r < s) {
            return {(ll) 9e18, (ll) -9e18};
        }
        if (l <= s && e <= r) {
            return seg[n];
        }
        int mid = (s + e) / 2;
        auto lval = query(n * 2, s, mid);
        auto rval = query(n * 2 + 1, mid + 1, e);
        ll mn = min(lval.mn, rval.mn);
        ll mx = max(lval.mx, rval.mx);
        return {mn, mx};
    }
};

int main() {
    ios::sync_with_stdio(false), cin.tie(nullptr), cout.tie(nullptr);
#ifndef ONLINE_JUDGE
    freopen("../input.txt", "r", stdin);
    freopen("../output.txt", "w", stdout);
#endif
    int n, m;
    cin >> n >> m;
    Seg seg(n);
    for (auto &i: seg.arr) {
        ll tmp;
        cin >> tmp;
        i = {tmp, tmp};
    }
    seg.init(1, 0, n - 1);
    while (m--) {
        ll a, b;
        cin >> a >> b;
        auto [mn, mx] = seg.segQuery(a - 1, b - 1);
        cout << mn << ' ' << mx << '\n';
    }
}
